File: python/src/tools/memory.py
"""Memory tool — manage knowledge: promote/demote scope, or delete entries."""

from ..store.metadata import MetadataStore
from ..store.scope import detect_project_root
from ..store.types import KnowledgeInput, KnowledgeSource
from ..store.vector import VectorStore


async def handle_memory(action: str, chunk_id: str) -> str:
    """Manage knowledge: promote/demote scope, or delete entries."""
    project_root = detect_project_root()

    if action == "delete":
        for scope in ("global", "project"):
            try:
                meta = MetadataStore(scope, project_root or None)  # type: ignore[arg-type]
                vector = VectorStore(scope, project_root or None)  # type: ignore[arg-type]

                if meta.delete(chunk_id):
                    vector.remove(chunk_id)
                    meta.close()
                    vector.close()
                    return f"Deleted knowledge entry {chunk_id} from {scope} scope."

                meta.close()
                vector.close()
            except Exception:
                continue

        return f"Knowledge entry {chunk_id} not found."

    # promote or demote
    from_scope = "project" if action == "promote" else "global"
    to_scope = "global" if action == "promote" else "project"

    if to_scope == "project" and not project_root:
        return "Cannot demote to project scope: no project root detected."

    try:
        from_meta = MetadataStore(from_scope, project_root or None)  # type: ignore[arg-type]
        from_vector = VectorStore(from_scope, project_root or None)  # type: ignore[arg-type]
        to_meta = MetadataStore(to_scope, project_root or None)  # type: ignore[arg-type]
        to_vector = VectorStore(to_scope, project_root or None)  # type: ignore[arg-type]

        chunk = from_meta.get_by_id(chunk_id)
        if not chunk:
            from_meta.close()
            from_vector.close()
            to_meta.close()
            to_vector.close()
            return f"Knowledge entry {chunk_id} not found in {from_scope} scope."

        input_ = KnowledgeInput(
            content=chunk.content,
            type=chunk.type,
            scope=to_scope,  # type: ignore[arg-type]
            project=chunk.project if to_scope == "project" else None,
            tags=chunk.tags,
            source=chunk.source,
            confidence=chunk.confidence,
        )
        inserted = to_meta.insert(input_)
        to_vector.index(inserted.id, inserted.content, inserted.tags)

        from_meta.delete(chunk_id)
        from_vector.remove(chunk_id)

        from_meta.close()
        from_vector.close()
        to_meta.close()
        to_vector.close()

        action_label = "Promoted" if action == "promote" else "Demoted"
        return (
            f"{action_label} knowledge entry.\n"
            f"{from_scope} \u2192 {to_scope}\n"
            f"New ID: {inserted.id}\n"
            f"Content: {chunk.content[:100]}..."
        )
    except Exception as err:
        return f"Error during {action}: {err}"

File: dart/lib/tools/memory.dart
import 'package:mcp_dart/mcp_dart.dart';
import '../store/metadata.dart';
import '../store/vector.dart';
import '../store/scope.dart';
import '../store/types.dart';

void registerMemoryTool(McpServer server) {
  server.registerTool(
    'memory',
    description: 'Manage knowledge: promote/demote scope, or delete entries',
    inputSchema: ToolInputSchema(
      properties: {
        'action': JsonSchema.string(
          enumValues: ['promote', 'demote', 'delete'],
          description:
              'promote: project\u2192global, demote: global\u2192project, delete: remove',
        ),
        'id': JsonSchema.string(
          description: 'Knowledge entry ID',
        ),
      },
      required: ['action', 'id'],
    ),
    callback: (args, extra) async {
      final action = args['action'] as String;
      final id = args['id'] as String;
      final projectRoot = detectProjectRoot();

      if (action == 'delete') {
        for (final scope in [KnowledgeScope.global, KnowledgeScope.project]) {
          try {
            final meta = MetadataStore(scope, projectRoot: projectRoot);
            final vector = VectorStore(scope, projectRoot: projectRoot);

            if (meta.delete(id)) {
              vector.remove(id);
              meta.close();
              vector.close();
              return CallToolResult(
                content: [
                  TextContent(
                    text:
                        'Deleted knowledge entry $id from ${scope.name} scope.',
                  ),
                ],
              );
            }

            meta.close();
            vector.close();
          } catch (_) {
            continue;
          }
        }

        return CallToolResult(
          content: [
            TextContent(text: 'Knowledge entry $id not found.'),
          ],
        );
      }

      // promote or demote
      final fromScope = action == 'promote'
          ? KnowledgeScope.project
          : KnowledgeScope.global;
      final toScope = action == 'promote'
          ? KnowledgeScope.global
          : KnowledgeScope.project;

      if (toScope == KnowledgeScope.project && projectRoot == null) {
        return CallToolResult(
          content: [
            TextContent(
              text:
                  'Cannot demote to project scope: no project root detected.',
            ),
          ],
        );
      }

      try {
        final fromMeta = MetadataStore(fromScope, projectRoot: projectRoot);
        final fromVector = VectorStore(fromScope, projectRoot: projectRoot);
        final toMeta = MetadataStore(toScope, projectRoot: projectRoot);
        final toVector = VectorStore(toScope, projectRoot: projectRoot);

        final chunk = fromMeta.getById(id);
        if (chunk == null) {
          fromMeta.close();
          fromVector.close();
          toMeta.close();
          toVector.close();
          return CallToolResult(
            content: [
              TextContent(
                text:
                    'Knowledge entry $id not found in ${fromScope.name} scope.',
              ),
            ],
          );
        }

        final input = KnowledgeInput(
          content: chunk.content,
          type: chunk.type,
          scope: toScope,
          project: toScope == KnowledgeScope.project ? chunk.project : null,
          tags: chunk.tags,
          source: chunk.source,
          confidence: chunk.confidence,
        );
        final inserted = toMeta.insert(input);
        toVector.index(inserted.id, inserted.content, inserted.tags);

        fromMeta.delete(id);
        fromVector.remove(id);

        fromMeta.close();
        fromVector.close();
        toMeta.close();
        toVector.close();

        final contentPreview = chunk.content.length > 100
            ? '${chunk.content.substring(0, 100)}...'
            : chunk.content;
        final actionLabel = action == 'promote' ? 'Promoted' : 'Demoted';

        return CallToolResult(
          content: [
            TextContent(
              text:
                  '$actionLabel knowledge entry.\n${fromScope.name} \u2192 ${toScope.name}\nNew ID: ${inserted.id}\nContent: $contentPreview',
            ),
          ],
        );
      } catch (err) {
        return CallToolResult(
          content: [
            TextContent(text: 'Error during $action: $err'),
          ],
        );
      }
    },
  );
}
